src/erlcloud_waf.erl

```diff
@@ -0,0 +1,1152 @@
+
+new(AccessKeyID, SecretAccessKey) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey
+      }.
+
+-spec new(string(), string(), string()) -> aws_config().
+
+new(AccessKeyID, SecretAccessKey, Host) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey,
+       waf_host=Host
+      }.
+
+-spec new(string(), string(), string(), non_neg_integer()) -> aws_config().
+
+new(AccessKeyID, SecretAccessKey, Host, Port) ->
+    #aws_config{
+       access_key_id=AccessKeyID,
+       secret_access_key=SecretAccessKey,
+       waf_host=Host,
+       waf_port=Port
+      }.
+
+-spec configure(string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey)),
+    ok.
+
+-spec configure(string(), string(), string()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host)),
+    ok.
+
+-spec configure(string(), string(), string(), non_neg_integer()) -> ok.
+
+configure(AccessKeyID, SecretAccessKey, Host, Port) ->
+    put(aws_config, new(AccessKeyID, SecretAccessKey, Host, Port)),
+    ok.
+
+default_config() ->
+    erlcloud_aws:default_config().
+
+
+-spec waf_result_fun/1 ::
+          (Request :: aws_request()) -> aws_request().
+waf_result_fun(#aws_request{response_type = ok} = Request) ->
+    Request;
+waf_result_fun(#aws_request{response_type = error,
+                                  error_type = aws,
+                                  response_status = Status} = Request) when Status >= 500 ->
+    Request#aws_request{should_retry = true};
+waf_result_fun(#aws_request{response_type = error, error_type = aws} = Request) ->
+    Request#aws_request{should_retry = false}.
+
+%%%------------------------------------------------------------------------------
+%%% AWS WAF API Functions
+%%%------------------------------------------------------------------------------
+
+%%%------------------------------------------------------------------------------
+%%% CreateByteMatchSet 
+%%
+%% http://docs.aws.amazon.com/waf/latest/APIReference/API_CreateByteMatchSet.html
+%%%------------------------------------------------------------------------------
+-spec create_byte_match_set/2 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary()) ->
+        waf_return_val().
+create_byte_match_set(ChangeToken, Name) ->
+    create_byte_match_set(ChangeToken, Name, default_config()).
+
+-spec create_byte_match_set/3 ::
+        (ChangeToken :: string() | binary(),
+         Name :: string() | binary(),
+         Config :: aws_config()) ->
+        waf_return_val().
+create_byte_match_set(ChangeToken, Name, Config) ->
+    Json = [{<<"ChangeToken">>, to_binary(ChangeToken)},
+            {<<"Name">>, to_binary(Name)}],
+    waf_request(Config, "AWSWAF_20150824.CreateByteMatchSet", Json).
```

Comment:
suggest hiding this API version as define and putting inside `waf_request()`. if new version of API will be released we will use ec2 module approach